ddtrace/contrib/molten/patch.py

```diff
@@ -0,0 +1,218 @@
+    if getattr(molten, '_datadog_patch', False):
+        setattr(molten, '_datadog_patch', False)
+
+        pin = Pin.get_from(molten)
+        pin.remove_from(molten)
+
+        unwrap(molten.BaseApp, '__init__')
+        unwrap(molten.App, '__call__')
+        unwrap(molten.Router, 'add_route')
+
+
+def trace_func(resource):
+    """Trace calls to function using provided resource name
+    """
+    @wrapt.function_wrapper
+    def _trace_func(wrapped, instance, args, kwargs):
+        pin = Pin.get_from(molten)
+
+        if not pin or not pin.enabled():
+            return wrapped(*args, **kwargs)
+
+        with pin.tracer.trace(func_name(wrapped), service=pin.service, resource=resource):
+            return wrapped(*args, **kwargs)
+
+    return _trace_func
+
+
+def trace_middleware(middleware):
+    """Trace calling of middleware function or object
+    """
+    @wrapt.function_wrapper
+    def _trace_middleware(wrapped, instance, args, kwargs):
+        pin = Pin.get_from(molten)
+        if not pin or not pin.enabled():
+            return wrapped(*args, **kwargs)
+
+        return trace_func(func_name(wrapped))(wrapped(*args, **kwargs))
+
+    return _trace_middleware(middleware)
+
+
+@wrapt.function_wrapper
+def patch_start_response(wrapped, instance, args, kwargs):
+    """Patch respond handling to set metadata
+    """
+    pin = Pin.get_from(molten)
+    if not pin or not pin.enabled():
+        return wrapped(*args, **kwargs)
+
+    span = pin.tracer.current_root_span()
+    if not span:
+        return wrapped(*args, **kwargs)
+
+    status, headers, exc_info = args
+    status_code, _, _ = status.partition(' ')
+    span.set_tag(http.STATUS_CODE, status_code)
+    return wrapped(*args, **kwargs)
+
+
+def patch_add_route(wrapped, instance, args, kwargs):
+    """Patch adding routes to trace route handler
+    """
+    def _wrap(route_like, prefix="", namespace=None):
+        # avoid patching non-Route, e.g. Include
+        if not isinstance(route_like, molten.Route):
+            return wrapped(*args, **kwargs)
+
+        resource = '{} {}'.format(
+            route_like.method,
+            route_like.template
+        )
+
+        # patch handler for route
+        route_like.handler = trace_func(resource)(route_like.handler)
+
+        return wrapped(route_like, prefix=prefix, namespace=namespace)
+
+    return _wrap(*args, **kwargs)
+
+
+def patch_app_call(wrapped, instance, args, kwargs):
+    """Patch wsgi interface for app
+    """
+    pin = Pin.get_from(molten)
+    if not pin or not pin.enabled():
+        return wrapped(*args, **kwargs)
+
+    # destruct arguments to wsgi app
+    environ, start_response = args
+    start_response = patch_start_response(start_response)
+    method = environ.get('REQUEST_METHOD')
+    path = environ.get('PATH_INFO')
+    resource = func_name(wrapped)
+
+    request = Request.from_environ(environ)
+    distributed_tracing = asbool(environ.get('DD_MOLTEN_DISTRIBUTED_TRACING')) or \
```

Comment:
I didn't realize before that you are trying to extract this from the `wsgi` environment, we don't want to do this, because that isn't something people will have control over. So it seems like it is added only for testing purposes, but we have other ways to enable this for testing.